src/prover_disk.hpp

```diff
@@ -167,33 +167,48 @@ class DiskProver {
+            for (int i = 0; i < p7_entries.size(); i++) {
+                tasks[i] = async(std::launch::async, [&, i] {
+                    uint64_t position = p7_entries[i];
+                    // Open the plot file for each task
+                    std::ifstream disk_file_(filename, std::ios::in | std::ios::binary);
+                    // This inner loop goes from table 6 to table 1, getting the two backpointers,
+                    // and following one of them.
+                    for (uint8_t table_index = 6; table_index > 1; table_index--) {
+                        uint128_t line_point = ReadLinePoint(disk_file_, table_index, position);
+
+                        auto xy = Encoding::LinePointToSquare(line_point);
+                        assert(xy.first >= xy.second);
+
+                        if (((last_5_bits >> (table_index - 2)) & 1) == 0) {
+                            position = xy.second;
+                        } else {
+                            position = xy.first;
+                        }
                     }
-                }
-                uint128_t new_line_point = ReadLinePoint(disk_file, 1, position);
-                auto x1x2 = Encoding::LinePointToSquare(new_line_point);
-
-                // The final two x values (which are stored in the same location) are hashed
-                std::vector<unsigned char> hash_input(32 + Util::ByteAlign(2 * k) / 8, 0);
-                memcpy(hash_input.data(), challenge, 32);
-                (LargeBits(x1x2.second, k) + LargeBits(x1x2.first, k))
-                    .ToBytes(hash_input.data() + 32);
-                std::vector<unsigned char> hash(picosha2::k_digest_size);
-                picosha2::hash256(hash_input.begin(), hash_input.end(), hash.begin(), hash.end());
-                qualities.emplace_back(hash.data(), 32, 256);
+                    uint128_t new_line_point = ReadLinePoint(disk_file_, 1, position);
+                    auto x1x2 = Encoding::LinePointToSquare(new_line_point);
+
+                    // The final two x values (which are stored in the same location) are hashed
+                    std::vector<unsigned char> hash_input(32 + Util::ByteAlign(2 * k) / 8, 0);
+                    memcpy(hash_input.data(), challenge, 32);
+                    (LargeBits(x1x2.second, k) + LargeBits(x1x2.first, k))
+                        .ToBytes(hash_input.data() + 32);
+                    std::vector<unsigned char> hash(picosha2::k_digest_size);
+                    picosha2::hash256(
+                        hash_input.begin(), hash_input.end(), hash.begin(), hash.end());
+                    // Wait for previous appending finish, order matters
+                    if (i > 0) {
+                        tasks[i - 1].wait();
+                    }
+                    qualities.emplace_back(hash.data(), 32, 256);
+                });
             }
+            // Waiting all read tasks finish
```

Comment:
this comment doesn't seem right. You're only waiting for the last task, all others have already completed, since they all wait for the one before it

Reply:
In line 206, that is the place that waiting for the other reading tasks. So it will only need to wait for the last one here.